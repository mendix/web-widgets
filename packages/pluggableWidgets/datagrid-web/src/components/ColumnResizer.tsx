import { useEventCallback } from "@mendix/widget-plugin-hooks/useEventCallback";
import { createElement, ReactElement, useCallback, useEffect, useRef, useState, MouseEvent, TouchEvent } from "react";

export interface ColumnResizerProps {
    minWidth?: number;
    setColumnWidth: (width: number) => void;
    onResizeEnds?: () => void;
    onResizeStart?: () => void;
}

export function ColumnResizer({
    minWidth = 50,
    setColumnWidth,
    onResizeEnds,
    onResizeStart
}: ColumnResizerProps): ReactElement {
    const [isResizing, setIsResizing] = useState(false);
    const [startPosition, setStartPosition] = useState(0);
    const [currentWidth, setCurrentWidth] = useState(0);
    const resizerReference = useRef<HTMLDivElement>(null);
    const onStart = useEventCallback(onResizeStart);

    const onStartDrag = useCallback(
        (e: TouchEvent<HTMLDivElement> & MouseEvent<HTMLDivElement>): void => {
            const mouseX = e.touches ? e.touches[0].screenX : e.screenX;
            setStartPosition(mouseX);
            setIsResizing(true);
            if (resizerReference.current) {
                const column = resizerReference.current.parentElement!;
                setCurrentWidth(column.clientWidth);
            }
            onStart();
        },
        [onStart]
    );
    const onEndDrag = useCallback((): void => {
        if (!isResizing) {
            return;
        }
        setIsResizing(false);
        setCurrentWidth(0);
        onResizeEnds?.();
    }, [onResizeEnds, isResizing]);
    const setColumnWidthStable = useEventCallback(setColumnWidth);
    const onMouseMove = useCallback(
        (e: TouchEvent & MouseEvent & Event): void => {
            if (!isResizing) {
                return;
            }
            const mouseX = e.touches ? e.touches[0].screenX : e.screenX;

            if (currentWidth) {
                const moveDifference = startPosition - mouseX;
                let newWidth = currentWidth - moveDifference;
                if (newWidth < minWidth) {
                    newWidth = minWidth;
                }
                setColumnWidthStable(newWidth);
            }
        },
        [isResizing, currentWidth, startPosition, minWidth, setColumnWidthStable]
    );

    useEffect(() => {
        document.addEventListener("mousemove", onMouseMove);
        document.addEventListener("mouseup", onEndDrag);

        document.addEventListener("touchmove", onMouseMove);
        document.addEventListener("touchend", onEndDrag);
        return () => {
            document.removeEventListener("mousemove", onMouseMove);
            document.removeEventListener("mouseup", onEndDrag);

            document.removeEventListener("touchmove", onMouseMove);
            document.removeEventListener("touchend", onEndDrag);
        };
    }, [onMouseMove, onEndDrag]);

    return (
        <div
            aria-hidden
            ref={resizerReference}
            className="column-resizer"
            onMouseDown={onStartDrag}
            onTouchStart={onStartDrag}
        >
            <div className="column-resizer-bar" />
        </div>
    );
}
