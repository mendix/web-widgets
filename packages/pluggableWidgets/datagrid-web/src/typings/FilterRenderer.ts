import { ReactElement, ReactNode } from "react";

export type FilterRenderer = (
    renderWrapper: (children: ReactNode) => ReactElement,
    columnIndex: number
) => ReactElement;
