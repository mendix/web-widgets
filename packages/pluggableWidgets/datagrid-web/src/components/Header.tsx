import {
    createElement,
    Dispatch,
    ReactElement,
    SetStateAction,
    DragEvent,
    DragEventHandler,
    KeyboardEvent,
    useCallback,
    HTMLAttributes
} from "react";
import classNames from "classnames";
import { FaLongArrowAltDown } from "./icons/FaLongArrowAltDown";
import { FaLongArrowAltUp } from "./icons/FaLongArrowAltUp";
import { FaArrowsAltV } from "./icons/FaArrowsAltV";
import { ColumnProperty } from "./Table";
import { ColumnResizerProps } from "./ColumnResizer";
import { SortingRule } from "../features/settings";

export interface HeaderProps {
    className?: string;
    column: ColumnProperty;
    sortable: boolean;
    resizable: boolean;
    filterable: boolean;
    draggable: boolean;
    dragOver: string;
    hidable: boolean;
    isDragging?: boolean;
    preview?: boolean;
    resizer: ReactElement<ColumnResizerProps>;
    setColumnOrder: (updater: string[]) => void;
    setDragOver: Dispatch<SetStateAction<string>>;
    setIsDragging: Dispatch<SetStateAction<boolean>>;
    setSortBy: Dispatch<SetStateAction<SortingRule[]>>;
    sortBy: SortingRule[];
    visibleColumns: ColumnProperty[];
}

export function Header(props: HeaderProps): ReactElement {
    const canSort = props.sortable && props.column.canSort;
    const canDrag = props.draggable && (props.column.canDrag ?? false);
    const draggableProps = useDraggable(
        canDrag,
        props.visibleColumns,
        props.setColumnOrder,
        props.setDragOver,
        props.setIsDragging
    );

    const [sortProperties] = props.sortBy;
    const isSorted = sortProperties && sortProperties.id === props.column.index.toString();
    const isSortedDesc = isSorted && sortProperties.desc;

    const sortIcon = canSort ? (
        isSorted ? (
            isSortedDesc ? (
                <FaLongArrowAltDown />
            ) : (
                <FaLongArrowAltUp />
            )
        ) : (
            <FaArrowsAltV />
        )
    ) : null;

    const caption = props.column.header.trim();

    const onSortBy = (): void => {
        /**
         * Always analyse previous values to predict the next
         * 1 - !isSorted turns to asc
         * 2 - isSortedDesc === false && isSorted turns to desc
         * 3 - isSortedDesc === true && isSorted turns to unsorted
         * If multisort is allowed in the future this should be changed to append instead of just return a new array
         */
        if (!isSorted) {
            props.setSortBy([{ id: props.column.index.toString(), desc: false }]);
        } else if (isSorted && !isSortedDesc) {
            props.setSortBy([{ id: props.column.index.toString(), desc: true }]);
        } else {
            props.setSortBy([]);
        }
    };

    const sortProps: HTMLAttributes<HTMLDivElement> = {
        onClick: onSortBy,
        onKeyDown: (e: KeyboardEvent<HTMLDivElement>) => {
            if (e.key === "Enter" || e.key === " ") {
                e.preventDefault();
                onSortBy();
            }
        },
        role: "button",
        tabIndex: 0
    };

    return (
        <div
            aria-sort={canSort ? (isSorted ? (isSortedDesc ? "descending" : "ascending") : "none") : undefined}
            className={classNames("th", {
                "hidden-column-preview": props.preview && props.hidable && props.column.hidden
            })}
            role="columnheader"
            style={!props.sortable || !props.column.canSort ? { cursor: "unset" } : undefined}
            title={caption}
        >
            <div
                className={classNames(
                    "column-container",
                    canDrag && props.column.id === props.dragOver ? "dragging" : ""
                )}
                id={props.column.id}
                {...draggableProps}
            >
                <div
                    className={classNames("column-header", canSort ? "clickable" : "", props.className)}
                    style={{ pointerEvents: props.isDragging ? "none" : undefined }}
                    {...(canSort ? sortProps : undefined)}
                    aria-label={canSort ? "sort " + caption : caption}
                >
                    <span>{caption.length > 0 ? caption : "\u00a0"}</span>
                    {sortIcon}
                </div>
                {props.filterable && props.column.customFilter ? props.column.customFilter : null}
            </div>
            {props.resizable && props.column.canResize && props.resizer}
        </div>
    );
}

function useDraggable(
    columnsDraggable: boolean,
    visibleColumns: ColumnProperty[],
    setColumnOrder: (updater: ((columnOrder: string[]) => string[]) | string[]) => void,
    setDragOver: Dispatch<SetStateAction<string>>,
    setIsDragging: Dispatch<SetStateAction<boolean>>
): {
    draggable?: boolean;
    onDragStart?: DragEventHandler;
    onDragOver?: DragEventHandler;
    onDrop?: DragEventHandler;
    onDragEnter?: DragEventHandler;
    onDragEnd?: DragEventHandler;
} {
    const handleDragStart = useCallback(
        (e: DragEvent<HTMLDivElement>): void => {
            setIsDragging(true);
            const { id } = e.target as HTMLDivElement;
            e.dataTransfer.setData("colDestination", id);
        },
        [setIsDragging]
    );

    const handleDragOver = useCallback((e: DragEvent<HTMLDivElement>): void => {
        e.preventDefault();
    }, []);

    const handleDragEnter = useCallback(
        (e: DragEvent<HTMLDivElement>): void => {
            const { id } = e.target as HTMLDivElement;
            const colDestination = e.dataTransfer.getData("colDestination");
            if (id !== colDestination) {
                setDragOver(id);
            }
        },
        [setDragOver]
    );

    const handleDragEnd = useCallback((): void => {
        setIsDragging(false);
        setDragOver("");
    }, [setDragOver, setIsDragging]);

    const handleOnDrop = useCallback(
        (e: DragEvent<HTMLDivElement>): void => {
            handleDragEnd();
            const { id: colOrigin } = e.target as HTMLDivElement;
            const colDestination = e.dataTransfer.getData("colDestination");

            const toIndex = visibleColumns.findIndex(col => col.id === colOrigin);
            const fromIndex = visibleColumns.findIndex(col => col.id === colDestination);

            if (toIndex !== fromIndex) {
                const newOrder = [...visibleColumns.map(column => column.id)];
                newOrder.splice(fromIndex, 1);
                newOrder.splice(toIndex, 0, colDestination);
                setColumnOrder(newOrder);
            }
        },
        [handleDragEnd, setColumnOrder, visibleColumns]
    );

    return columnsDraggable
        ? {
              draggable: true,
              onDragStart: handleDragStart,
              onDragOver: handleDragOver,
              onDrop: handleOnDrop,
              onDragEnter: handleDragEnter,
              onDragEnd: handleDragEnd
          }
        : {};
}
