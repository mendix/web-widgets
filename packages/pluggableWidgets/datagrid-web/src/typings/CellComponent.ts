import { ObjectItem } from "mendix";
import { ReactElement, ReactNode } from "react";
import { GridColumn } from "./GridColumn";
import { ElementProps } from "@mendix/widget-plugin-grid/event-switch/base";

export interface EventsController {
    getProps(item: ObjectItem): ElementProps<HTMLDivElement>;
}

export interface CellComponentProps<C extends GridColumn> {
    children?: ReactNode;
    className?: string;
    column: C;
    item: ObjectItem;
    key?: string | number;
    rowIndex: number;
    columnIndex?: number;
    clickable?: boolean;
    preview?: boolean;
    eventsController: EventsController;
}

export type CellComponent<C extends GridColumn> = (props: CellComponentProps<C>) => ReactElement;
