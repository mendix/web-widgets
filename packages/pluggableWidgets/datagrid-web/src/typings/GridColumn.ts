import { ObjectItem } from "mendix";
import { ReactElement } from "react";
import { AlignmentEnum, WidthEnum } from "../../typings/DatagridProps";

export type ColumnId = string & { __columnIdTag: never };

/**
 * A generic column type for data grid.
 */
export interface GridColumn {
    alignment: AlignmentEnum;
    canDrag: boolean;
    canHide: boolean;
    canResize: boolean;
    canSort: boolean;
    columnClass(item: ObjectItem): string | undefined;
    columnId: ColumnId;
    columnNumber: number;
    header: string;
    initiallyHidden: boolean;
    renderCellContent: (item: ObjectItem) => ReactElement;
    visible: boolean;
    weight: number;
    width: WidthEnum;
    wrapText: boolean;
}
