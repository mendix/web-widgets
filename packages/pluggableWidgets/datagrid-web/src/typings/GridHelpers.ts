import { EventsController } from "./CellComponent";
import { FocusTargetController } from "@mendix/widget-plugin-grid/keyboard-navigation/FocusTargetController";
import { SelectActionHelper } from "../helpers/SelectActionHelper";
import { SelectionHelper } from "@mendix/widget-plugin-grid/selection";

export type GridHelpers = {
    readonly cellEventsController: EventsController;
    readonly checkboxEventsController: EventsController;
    readonly focusController: FocusTargetController;
    readonly selectActionHelper: SelectActionHelper;
    readonly selectionHelper: SelectionHelper | undefined;
};
