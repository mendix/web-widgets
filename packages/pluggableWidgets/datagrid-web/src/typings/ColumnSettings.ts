import { ColumnId } from "./GridColumn";

export interface ColumnSettings {
    columnId: ColumnId;
    size: number | undefined;
    hidden: boolean;
    filterSettings: unknown;
}
