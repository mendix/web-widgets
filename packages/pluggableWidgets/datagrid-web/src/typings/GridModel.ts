import { ListValue } from "mendix";
import { FilterCondition } from "mendix/filters";
import { ColumnId, GridColumn } from "./GridColumn";

export interface ColumnWidthConfig {
    [columnId: ColumnId]: number | undefined;
}

export type SortInstruction = ListValue["sortOrder"] extends Array<infer T> ? T : never;

export type SortDirection = SortInstruction[1];

export type SortRule = [columnId: ColumnId, dir: SortDirection];

export type Columns = GridColumn[];

export type Filter = FilterCondition | undefined;

export type Hidden = Set<ColumnId>;

export type ColumnOrder = ColumnId[];

export type SortOrder = SortRule[];

interface BaseState {
    columnOrder: ColumnOrder;
    hidden: Hidden;
    size: ColumnWidthConfig;
    sortOrder: SortOrder;
}

export interface StorableState extends BaseState {
    columns: Columns;
    name: string;
    settingsHash: string;
}

export interface InitParams extends BaseState {
    filter: Filter;
}

export interface State extends BaseState {
    allColumns: Columns;
    availableColumns: Columns;
    visibleColumns: Columns;
    filter: Filter;
}

export type Actions = {
    toggleHidden: (id: ColumnId) => void;
    sortBy: (id: ColumnId) => void;
    swap: (arg: [a: ColumnId, b: ColumnId]) => void;
    resize: (arg: [id: ColumnId, size: number]) => void;
    setFilter: (arg: Filter) => void;
};
