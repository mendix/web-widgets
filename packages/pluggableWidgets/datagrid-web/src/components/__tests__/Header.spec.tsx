import { render, shallow } from "enzyme";
import { createElement } from "react";
import { Header, HeaderProps } from "../Header";
import { ColumnResizer } from "../ColumnResizer";
import { GridColumn } from "../../typings/GridColumn";

describe("Header", () => {
    it("renders the structure correctly", () => {
        const component = render(<Header {...mockHeaderProps()} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when sortable", () => {
        const props = mockHeaderProps();
        props.column.canSort = true;
        props.sortable = true;

        const component = render(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when resizable", () => {
        const props = mockHeaderProps();
        props.column.canResize = true;
        props.resizable = true;

        const component = render(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when draggable", () => {
        const props = mockHeaderProps();
        props.column.canDrag = true;
        props.draggable = true;

        const component = render(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when filterable with no custom filter", () => {
        const props = mockHeaderProps();
        props.filterable = true;

        const component = render(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when filterable with custom filter", () => {
        const props = mockHeaderProps();
        const filterWidget = (
            <div>
                <label>Date picker filter</label>
                <input type="date" />
            </div>
        );
        props.filterable = true;

        const component = render(<Header {...props} filterWidget={filterWidget} />);

        expect(component).toMatchSnapshot();
    });

    it("calls setSortBy store function with correct parameters when sortable", () => {
        const column = {
            columnId: "0",
            columnNumber: 0,
            header: "My sortable column",
            canSort: true
        } as any;
        const mockedFunction = jest.fn();
        const component = shallow(
            <Header {...mockHeaderProps()} column={column} sortable setSortBy={mockedFunction} />
        );

        const clickableRegion = component.find(".column-header");

        expect(clickableRegion).toHaveLength(1);

        clickableRegion.simulate("click");
        expect(mockedFunction).toHaveBeenLastCalledWith("0");
    });

    it("renders the structure correctly when filterable with custom classes", () => {
        const props = mockHeaderProps();
        props.filterable = true;

        const component = render(<Header {...props} className="my-custom-class" />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when is hidden and preview", () => {
        const props = mockHeaderProps();
        props.column.initiallyHidden = true;
        props.hidable = true;
        props.preview = true;

        const component = render(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when value is empty", () => {
        const props = mockHeaderProps();
        props.column.header = " ";

        const component = render(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });
});

function mockHeaderProps(): HeaderProps {
    return {
        gridId: "dg1",
        column: {
            columnId: "dg1-column0",
            columnNumber: 0,
            header: "Test"
        } as GridColumn,
        draggable: false,
        dragOver: undefined,
        filterable: false,
        filterWidget: undefined,
        hidable: false,
        resizable: false,
        resizer: <ColumnResizer setColumnWidth={jest.fn()} />,
        sortable: false,
        swapColumns: jest.fn(),
        setDragOver: jest.fn(),
        visibleColumns: [],
        setSortBy: jest.fn(),
        setIsDragging: jest.fn(),
        sortRule: undefined
    };
}
