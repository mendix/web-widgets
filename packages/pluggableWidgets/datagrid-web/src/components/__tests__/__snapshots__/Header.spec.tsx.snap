// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders the structure correctly 1`] = `
<div
  class="th"
  role="columnheader"
  style="cursor:unset"
  title="Test"
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label="Test"
      class="column-header"
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when draggable 1`] = `
<div
  class="th"
  role="columnheader"
  style="cursor:unset"
  title="Test"
>
  <div
    class="column-container"
    draggable="true"
    id="dg1-column1"
  >
    <div
      aria-label="Test"
      class="column-header"
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when filterable with custom classes 1`] = `
<div
  class="th"
  role="columnheader"
  style="cursor:unset"
  title="Test"
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label="Test"
      class="column-header my-custom-class"
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when filterable with custom filter 1`] = `
<div
  class="th"
  role="columnheader"
  style="cursor:unset"
  title="Test"
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label="Test"
      class="column-header"
    >
      <span>
        Test
      </span>
    </div>
    <div>
      <label>
        Date picker filter
      </label>
      <input
        type="date"
      />
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when filterable with no custom filter 1`] = `
<div
  class="th"
  role="columnheader"
  style="cursor:unset"
  title="Test"
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label="Test"
      class="column-header"
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when is hidden and preview 1`] = `
<div
  class="th hidden-column-preview"
  role="columnheader"
  style="cursor:unset"
  title="Test"
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label="Test"
      class="column-header"
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when resizable 1`] = `
<div
  class="th"
  role="columnheader"
  style="cursor:unset"
  title="Test"
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label="Test"
      class="column-header"
    >
      <span>
        Test
      </span>
    </div>
  </div>
  <div
    aria-hidden="true"
    class="column-resizer"
  >
    <div
      class="column-resizer-bar"
    />
  </div>
</div>
`;

exports[`Header renders the structure correctly when sortable 1`] = `
<div
  aria-sort="none"
  class="th"
  role="columnheader"
  title="Test"
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label="sort Test"
      class="column-header clickable"
      role="button"
      tabindex="0"
    >
      <span>
        Test
      </span>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrows-alt-v fa-w-8"
        data-icon="arrows-alt-v"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when value is empty 1`] = `
<div
  class="th"
  role="columnheader"
  style="cursor:unset"
  title=""
>
  <div
    class="column-container"
    id="dg1-column1"
  >
    <div
      aria-label=""
      class="column-header"
    >
      <span>
        Â 
      </span>
    </div>
  </div>
</div>
`;
